ARROW-8494: [C++][Parquet] Full support for reading mixed list and structs

emkornfield opened the pull request:
Also: ARROW-9810 (generalize rep/def level conversion to list lengths/bitmaps)

This adds helper methods for reconstructing all necessary metadata
for arrow types.  For now this doesn't handle null_slot_usage (i.e.
children of FixedSizeList), it throws exceptions when nulls are
encountered in this case.  It uses there for generic reconstruction.

The unit tests demonstrate how to use the helper methods in combination
with LevelInfo (generated from parquet/arrow/schema.h) to reconstruct
the metadata.  The arrow reader.cc is now rewritten to use these method.

- Refactors necessary APIs to use LevelInfo and makes use of them in
  column_reader
- Adds implementations for reconstructing list validity bitmaps
  (uses rep/def levels)
- Adds implementations for reconstruction list lengths
 (uses rep/def levels.).
- Adds dynamic dispatch for level comparison algorithms for AVX2
  and BMI2.
- Adds a pextract alternative that uses BitRunReader that can be
  used as a fallback.
- Fixes some bugs in detailed reconstruction to array tests.

emkornfield commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,169 @@
+      // filtered out above).
+      if (lengths != nullptr) {
+        ++lengths;
+        *lengths = (def_levels[x] >= level_info.def_level) ? 1 : 0;
```
we should consider doing the sume of the previous element here.  Originally I did not because I thought at some point getting raw lengths would make it easier to handled chunked_arrays in reader.cc but I think that case is esoteric enough that removing the need to touch this data twice will be better.

emkornfield commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -692,14 +692,6 @@ def test_pandas_can_write_nested_data(tempdir):
     # This succeeds under V2
     _write_table(arrow_table, imos)
 
-    # Under V1 it fails.
```
this was meant for my other PR< I willl revert it.

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,190 @@
+        *lengths += 1;
+      }
+    } else {
+      // current_rep < list rep_level i.e. start of a list (ancenstor empty lists are
```
"ancestor". Do you have a buggy automatic spell checker? :-)

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,190 @@
-#endif  // ARROW_HAVE_BMI2
+    if (rep_levels[x] == level_info.rep_level) {
+      // A continuation of an existing list.
+      if (lengths != nullptr) {
```
Is there a reason we allow passing `nullptr` for `lengths` (which should be `offsets`?)

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,190 @@
+      // filtered out above).
+      if (lengths != nullptr) {
+        ++lengths;
+        // Use cumulative lengths because this is what the more common
```
I don't understand the "more common" comment. Do some Arrow list types have non-cumulative lengths?

pitrou commented on cpp/src/parquet/level_conversion_inc.h:
```diff
@@ -0,0 +1,146 @@
+    // repeated_ancenstor_def_level
+    uint64_t present_bitmap = internal::GreaterThanBitmap(
+        def_levels, batch_size, level_info.repeated_ancestor_def_level - 1);
+    uint64_t selected_bits = ExtractBits(defined_bitmap, present_bitmap);
```
FTR, I think that if BMI isn't available, you can still use a batch size of 5 or 6 bits and use a fast lookup table for ExtractBits (rather than the probably slow emulation code).

pitrou replied:
Yeah, we can think about that for another PR anyway. Will try to run benchmarks.

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,43 +137,49 @@ struct PARQUET_EXPORT LevelInfo {
+  int64_t values_read = 0;
+  /// The number of nulls encountered.
+  int64_t null_count = 0;
+  // The validity bitmp to populate. Can only be null
```
Input only?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,43 +137,49 @@ struct PARQUET_EXPORT LevelInfo {
+};
 
+/// Converts def_levels to validity bitmaps for non-list arrays.
+void PARQUET_EXPORT DefinitionLevelsToBitmap(const int16_t* def_levels,
```
The variations in terminology confuse me a lot. There's "DefinitionLevelsToBitmap" and "ConvertDefRepLevelsToBitmap". Can we settle on a clear naming scheme, so that it's apparent what the actual differences are?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,43 +137,49 @@ struct PARQUET_EXPORT LevelInfo {
+    LevelInfo level_info, ValidityBitmapInputOutput* output,
+    ::arrow::util::variant<int32_t*, int64_t*> lengths);
+
+/// Reconstructs a validity bitmap for a struct that has nested children.
```
Or you mean "nested list children and/or ancestor lists"? Nested struct children are indifferent, AFAICT.

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,43 +137,49 @@ struct PARQUET_EXPORT LevelInfo {
+uint64_t PARQUET_EXPORT RunBasedExtract(uint64_t bitmap, uint64_t selection);
+
+#if defined(ARROW_HAVE_RUNTIME_BMI2)
+void PARQUET_EXPORT DefinitionLevelsToBitmapBmi2WithRepeatedParent(
```
Hmm... does this need to be exposed here? Again, it's a bit confusing to have instruction set specific implementation details at this place.

pitrou reviewed (commented):
Ok, thanks a lot for this PR. I think I am understanding the implementation (I skipped `parquet/arrow/reader.cc` for now, though). Some of the implementation details are still confusing me a bit. In any case, here are some comments.

pitrou commented:
Are there any benchmarks worth running here?

emkornfield commented:
> Ok, thanks a lot for this PR. I think I am understanding the implementation (I skipped parquet/arrow/reader.cc for now, though). Some of the implementation details are still confusing me a bit. In any case, here are some comments.

Please let me know if there is more confusion, I will attempt to add clarifying comments.  I think I addressed all your comments except for some in level_conversion_test.cc I'll address those tomorrow (I assume there will be more comments in reader.cc as well).

> Are there any benchmarks worth running here?

parquet-level-conversion-benchmark
parquet-arrow-reader-writer-benchmark (this won't cover the nested cases though)  There is an open JIRA under ARROW-1644 to add benchmarks for nested cases @npr mentioned there might be some example datasets that we wanted to try this on.

pitrou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -50,7 +50,7 @@ if(ARROW_CPU_FLAG STREQUAL "x86")
     set(ARROW_AVX512_FLAG "-march=skylake-avx512 -mbmi2")
     # Append the avx2/avx512 subset option also, fix issue ARROW-9877 for homebrew-cpp
-    set(ARROW_AVX2_FLAG "${ARROW_AVX2_FLAG} -mavx2")
+    set(ARROW_AVX2_FLAG "${ARROW_AVX2_FLAG} -mavx2 -mbmi2")
```
What is your opinion on this @emkornfield ?

emkornfield replied:
did you mean to tag someone else here?  Or do I get the final say :)

pitrou replied:
No, I though you hadn't seen this comment :-)
If there are CPUs supporting one but not the other, then we must use separate detection. We don't want binaries to crash on some CPUs, even if they are uncommon.

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -108,22 +114,245 @@ TEST(GreaterThanBitmap, GeneratesExpectedBitmasks) {
+  this->test_data_.rep_levels_ = std::vector<int16_t>{0};
+
+  std::vector<typename TypeParam::ListLengthType> lengths(
+      2, std::numeric_limits<typename TypeParam::ListLengthType>::max());
```
Ah, I understand, offsets will be computed by adding to the first one... can you add a comment?

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -108,22 +114,245 @@ TEST(GreaterThanBitmap, GeneratesExpectedBitmasks) {
+  uint64_t validity_output;
+  ValidityBitmapInputOutput validity_io;
+  validity_io.values_read_upper_bound = 4;
+  validity_io.valid_bits = reinterpret_cast<uint8_t*>(&validity_output);
```
Ping here.

pitrou commented on cpp/src/parquet/level_conversion_test.cc:
```diff
@@ -108,22 +114,245 @@ TEST(GreaterThanBitmap, GeneratesExpectedBitmasks) {
+  EXPECT_EQ(validity_io.values_read, 6);
+  EXPECT_EQ(validity_io.null_count, 0);
+  EXPECT_EQ(BitmapToString(validity_io.valid_bits, /*length=*/6), "111111");
+}
```
Not sure what you think about this?

emkornfield replied:
sorry, I didn't respond directly here.  I'll refactor the tests tonight.

emkornfield replied:
For this, I think if we have good coverage at the reconstruction level it is likely better.  I refactored the tests so if we do identify bugs in the future it will be easy to add.   

My main motivation here is I missed a couple of bugs by focusing heavily on the analogous code and not end-to-end (this was partially hampered by not having the read side done).

pitrou commented on cpp/src/parquet/level_conversion_inc.h:
```diff
@@ -0,0 +1,146 @@
+  int64_t selected_bits = 0;
+  while (run.length != 0) {
+    if (run.set) {
+      new_bitmap |= (bitmap & ::arrow::BitUtil::LeastSignficantBitMask(run.length))
```
Thanks for the JIRA.

pitrou commented on cpp/src/parquet/level_conversion_inc.h:
```diff
@@ -0,0 +1,139 @@
+}
+
+inline uint64_t ExtractBits(uint64_t bitmap, uint64_t select_bitmap) {
+#if defined(ARROW_HAVE_BMI2) && !defined(__MINGW32__)
```
Comment why mingw is left out?

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,44 +135,65 @@ struct PARQUET_EXPORT LevelInfo {
+//  Converts def_levels to validity bitmaps for non-list arrays and structs that have
+//  at least one member that is not a list and has no list descendents.
+//  For lists use DefRepLevelsToList and structs where all descendants contain
+//  a list use DefRepLevelsToBitmap.
```
Fun :-)

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -132,44 +135,65 @@ struct PARQUET_EXPORT LevelInfo {
+// the next offset.  See documentation onf DefLevelsToBitmap for when to use this
+// method vs the other ones in this file for reconstruction.
+//
+// Offsets must be size to 1 + values_read_upper_bound.
```
Great comments, thank you.

pitrou commented on cpp/src/parquet/level_conversion.h:
```diff
@@ -19,6 +19,7 @@
 
 #include <cstdint>
 
+#include "arrow/util/bitmap.h"
```
Looks like this include is not used after all?

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,205 @@
+            : standard::DefLevelsToBitmapSimd</*has_repeated_parent=*/true>;
+    fn(def_levels, num_def_levels, level_info, output);
+#else
+    // This indicates we are likely on a big-endian platformat which don't have a
```
"platform"

But the comment is mistaken: ARM is little-endian most of the time (technically it supports both, but Linux runs it in little-endian mode AFAIK).

Also, I don't understand why `DefLevelsToBitmapScalar` is preferred here but `DefLevelsToBitmapSimd` is preferred below? Don't the same arguments apply?

pitrou commented on cpp/src/parquet/level_conversion.cc:
```diff
@@ -18,176 +18,190 @@
+      /*start_offset=*/output->valid_bits_offset,
+      /*length=*/num_def_levels);
+  for (int x = 0; x < num_def_levels; x++) {
+    if (def_levels[x] < level_info.repeated_ancestor_def_level) {
```
Not sure if the second question makes sense?

pitrou replied:
Ok, thanks.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+    ::parquet::internal::DefRepLevelsToList(def_levels, rep_levels, num_levels,
+                                            level_info_, &validity_io, length_data);
+    END_PARQUET_CATCH_EXCEPTIONS
+    // We might return less then the requested slot (i.e. reaching an end of a file)
```
Or is `number_of_slots` just an upper bound?

emkornfield replied:
removed.  it is an upper bound. renamed it. and removed comment.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayData> item_chunk, ChunksToSingle(**out));
+
+    std::vector<std::shared_ptr<Buffer>> buffers{
+        validity_io.null_count > 0 ? validity_buffer : std::shared_ptr<Buffer>(),
```
Can just write `nullptr`, I think.

emkornfield replied:
yes, I thought this was causing problems with type inference at some pont.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+        std::vector<std::shared_ptr<ArrayData>>{item_chunk}, validity_io.null_count);
+
+    std::shared_ptr<Array> result = ::arrow::MakeArray(data);
+    RETURN_NOT_OK(result->Validate());
```
I'm not sure we want to validate here. Let the user (and the tests) do it?

pitrou replied:
(the tests should use ValidateFull, anyway)

emkornfield replied:
I move this to above to avoid recursively calling things multiple times, but i think we should be validating at least for structs (and not validating full) since rep/def level information could be inconsistent within them.  It felt easier to call validate (and not too expensive) then writing custom logic for this.

I'm open to removing them, but it feels like there should be a contract here that someplace in this code for an underlying library we validate consistency.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+        children_(std::move(children)) {
+    // There could be a mix of children some might be repeated some might not be.
+    // If possible use one that isn't since that will be guaranteed to have the least
+    // number of rep/def levels.
```
You mean "of rep levels"? You could have arbitrarily nested structs with a lot of dep levels?

emkornfield replied:
rephrased a little bit.  There is always an equal number of repetition and definition levels for any particular leaf.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-    // Empty struct
     *length = 0;
-    return Status::OK();
+    return Status::Invalid("StructReader had no childre");
```
"children"

emkornfield replied:
done.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-  if (!found_nullable_child) {
+  if (*data == nullptr) {
+    // Only happens if there are actually 0 rows available.
     *data = nullptr;
```
Isn't this redundant with your `if` condition above? Also, why does `length` need to be filled out explicitly below?
(doesn't `def_rep_level_child_->GetDefLevels` do it?).

emkornfield replied:
Yes I think it is, removed.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-    }
+  if (data == nullptr) {
+    // Only happens if there are actually 0 rows available.
+    *data = nullptr;
```
You're dereferencing a null pointer (see `if` condition above).

emkornfield replied:
yeah, removed.  we shouldn't need this.

pitrou commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+    DefLevelsToBitmap(def_levels, num_levels, level_info_, &validity_io);
+  }
+
+  // Ensure all values are initialized.
```
Why? Shouldn't you resize the buffer instead?

emkornfield replied:
I didn't think about using Resizable buffers.  changed all of the places that were filled to it.

pitrou reviewed (commented):
Some comments on `reader.cc` now.

pitrou commented:
`parquet-level-conversion-benchmark`:
* before:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         966 ns          966 ns       714328 bytes_per_second=1.9753G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1626 ns         1626 ns       426358 bytes_per_second=1.173G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       1850 ns         1850 ns       367809 bytes_per_second=1055.83M/s
```
* after:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         560 ns          560 ns      1239730 bytes_per_second=3.40515G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent         562 ns          562 ns      1244256 bytes_per_second=3.39684G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent        562 ns          562 ns      1244691 bytes_per_second=3.39515G/s
```

Impressive!

pitrou commented:
No changes on `parquet-arrow-reader-writer-benchmark`. I suspect it doesn't trigger any of the updated code?

emkornfield commented:
> No changes on parquet-arrow-reader-writer-benchmark. I suspect it doesn't trigger any of the updated code?

No, I don't think so.  Since we were already using SIMD for non-nested types.

emkornfield commented:
@pitrou unfortunately, I was missing an "info.rep_level = 1;" in the benchmark, so it likely not as impressive on AMD, would you mind running again? (working on addressing the rest of the feedback.

emkornfield commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+    // ensure we've set all the bits here.
+    if (validity_io.values_read < number_of_slots) {
+      // + 1  because arrays lengths are values + 1
+      std::fill(length_data + validity_io.values_read + 1,
```
changed to resize.

emkornfield commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+                       ctx_->pool));
+    // ensure zero initialization in case we have reached a zero length list (and
+    // because first entry is always zero).
+    IndexType* length_data = reinterpret_cast<IndexType*>(lengths_buffer->mutable_data());
```
done.

emkornfield reviewed (commented):
still need to refactor level_conversion_test but I think i forgot to respond to the last review here.

emkornfield commented:
@pitrou I think I responded to all review comments at this point, apologies if I missed something.  level_conversion_test.cc is now refactored to a point where duplicate code I think adds to test understandability but there is still some redundancy.  Also, please see my note about the level_conversion_benchmark having a bug in it on your prior run.

nealrichardson commented:
I think the macOS failure is fixed by #8196 but the Appveyor failure looks legit:

```
C:/projects/arrow/cpp/src/parquet/level_conversion_test.cc(295): error C2220: warning treated as error - no 'object' file generated
C:/projects/arrow/cpp/src/parquet/level_conversion_test.cc(278): note: while compiling class template member function 'void parquet::internal::NestedListTest_SimpleLongList_Test<T>::TestBody(void)'
        with
        [
            T=Type
        ]
googletest_ep-prefix\include\gtest/internal/gtest-internal.h(665): note: see reference to class template instantiation 'parquet::internal::NestedListTest_SimpleLongList_Test<T>' being compiled
        with
        [
            T=Type
        ]
googletest_ep-prefix\include\gtest/internal/gtest-internal.h(657): note: while compiling class template member function 'bool testing::internal::TypeParameterizedTest<parquet::internal::NestedListTest,testing::internal::TemplateSel<parquet::internal::NestedListTest_SimpleLongList_Test>,parquet::internal::gtest_type_params_NestedListTest_>::Register(const char *,const testing::internal::CodeLocation &,const char *,const char *,int,const std::vector<std::string,std::allocator<_Ty>> &)'
        with
        [
            _Ty=std::string
        ]
C:/projects/arrow/cpp/src/parquet/level_conversion_test.cc(278): note: see reference to function template instantiation 'bool testing::internal::TypeParameterizedTest<parquet::internal::NestedListTest,testing::internal::TemplateSel<parquet::internal::NestedListTest_SimpleLongList_Test>,parquet::internal::gtest_type_params_NestedListTest_>::Register(const char *,const testing::internal::CodeLocation &,const char *,const char *,int,const std::vector<std::string,std::allocator<_Ty>> &)' being compiled
        with
        [
            _Ty=std::string
        ]
C:/projects/arrow/cpp/src/parquet/level_conversion_test.cc(278): note: see reference to class template instantiation 'testing::internal::TypeParameterizedTest<parquet::internal::NestedListTest,testing::internal::TemplateSel<parquet::internal::NestedListTest_SimpleLongList_Test>,parquet::internal::gtest_type_params_NestedListTest_>' being compiled
C:/projects/arrow/cpp/src/parquet/level_conversion_test.cc(295): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
[212/245] Linking CXX shared library release\parquet.dll
   Creating library release\parquet.lib and object release\parquet.exp
[213/245] Building CXX object src\parquet\CMakeFiles\parquet-arrow-test.dir\Unity\unity_0_cxx.cxx.obj
ninja: build stopped: subcommand failed.
```

nealrichardson commented:
👍 we've reduced the failures to Flight (i.e. surely unrelated) issues.

pitrou commented:
I'll take a look again on Monday, if that's ok with you.

emkornfield commented:
SGTM.  If i have time I might get one or two CLs out based on this one but I can rebase afterwards.

pitrou commented on cpp/src/parquet/CMakeLists.txt:
```diff
@@ -202,6 +203,19 @@ set(PARQUET_SRCS
     types.cc)
 
+if(CXX_SUPPORTS_AVX2)
+  # AVX2 is used as a proxy for BMI2.
```
According to https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html , it seems that the following CPUs may have support for AVX2 but not BMI2: Intel Alder Lake, Via Eden. This is obviously an extremely minor fraction of the CPU market, but I wouldn't be surprised if some day we get some bug reports because of it (or worse, people will decide that Arrow isn't reliable because it crashes for some of their customers).

emkornfield replied:
@pitrou I added a comment below about why I think the current config should be OK (and why people should not copy the flags).  At this point, I believe the only place BMI2 is enabled is on level_conversion_bmi2.cc  which at this point should always be guarded by runtime dispatch that specifically checks for BMI2.  Thoughts?

pitrou replied:
Ahah, sorry. Indeed, for some reason I had misread that `-mbmi2` would be applied to all source files. I agree it's ok, then.

pitrou commented:
Running `parquet-level-conversion-benchmark` again, results are much less good:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         878 ns          878 ns       816435 bytes_per_second=2.17265G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent         967 ns          966 ns       713185 bytes_per_second=1.9736G/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       3258 ns         3258 ns       214817 bytes_per_second=599.563M/s
```

pitrou commented:
Things are a bit more balanced if the scalar version is used:
```
BM_DefinitionLevelsToBitmapRepeatedAllMissing         660 ns          660 ns      1019688 bytes_per_second=2.8884G/s
BM_DefinitionLevelsToBitmapRepeatedAllPresent        1946 ns         1946 ns       358335 bytes_per_second=1003.9M/s
BM_DefinitionLevelsToBitmapRepeatedMostPresent       2024 ns         2023 ns       346078 bytes_per_second=965.229M/s
```

pitrou commented:
Just for the record, apart from FixedSizeList, is there anything remaining for full nested Parquet -> Arrow reading?

pitrou commented:
Other than that, I see a ~20% improvement on `BM_ReadStructColumn` and `BM_ReadListColumn`

pitrou commented:
I have no remaining concern over the code other than the AVX2 / BMI2 split. Congratulations for this PR, this is really a huge improvement!

That said, I seem to get a test error on the Python side (pasted below). Let's see if it reproduces on CI:
```
Traceback (most recent call last):
  File "/home/antoine/arrow/dev/python/pyarrow/tests/test_parquet.py", line 700, in test_pandas_can_write_nested_data
    _write_table(arrow_table, imos)
  File "/home/antoine/miniconda3/envs/pyarrow/lib/python3.7/site-packages/_pytest/python_api.py", line 747, in __exit__
    fail(self.message)
  File "/home/antoine/miniconda3/envs/pyarrow/lib/python3.7/site-packages/_pytest/outcomes.py", line 128, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: DID NOT RAISE <class 'ValueError'>
```

(beware: I rebased your branch on git master)

emkornfield commented:
> I have no remaining concern over the code other than the AVX2 / BMI2 split. Congratulations for this PR, this is really a huge improvement!

@pitrou thank you for the thoughtful review.  Let me know if you still have issues with the AVX/BMI2 after I added the comment (perhaps I didn't revert some compilation) or my analysis is wrong.  I think the BMI2 check will be difficult/impossible at compile time for windows, so I'm not sure if it is worth the effort  on linux.

I also removed the failing test for parquet (which I should have removed in a prior PR, its strange it showed up again).

emkornfield commented:
> Just for the record, apart from FixedSizeList, is there anything remaining for full nested Parquet -> Arrow reading?

We need to support LargeList, and Map which should be smaller change (I'm working on a PR) at the schema level inference.  There are a few other JIRAs still open about benchmarking and randomized testing,   Past that, there are some open JIRAs about performance improvements:
*  Computing all all offsets/bitmaps together (the JIRA is about non-vectorized).  I would expect that for deeply nested structures containing lists this would start to show performance improvements.
*  Using bitmap based code that was removed from this.  For non-list types I think it can be a big performance win on all platforms and a win at least for shallowly nested lists I expect it to be better for native Inte.

There is also an unrelated bug on the write side https://github.com/apache/arrow/pull/8219 which I asked for @wesm to review (it is based on some changes in this PR).

pitrou commented:
Do we also need ad hoc nested tests as a separate JIRA / PR? Randomized testing is nice to find corner cases, but it's always easier to diagnose hand-written test cases :-)

pitrou commented:
Also we only have one-level nested benchmarks for now, I suppose we should add a bit more (two-level nesting may be enough).

emkornfield commented:
> Do we also need ad hoc nested tests as a separate JIRA / PR? Randomized testing is nice to find corner cases, but it's always easier to diagnose hand-written test cases :-)

I think the internal tests you wrote have pretty good coverage.  After #8219 is merged I was planning on make some of the one way (the ones I wrote for write and the ones your wrote for read) fully round-trip.  If you think there are gaps, by all means we should add tests.

emkornfield commented:
> Also we only have one-level nested benchmarks for now, I suppose we should add a bit more (two-level nesting may be enough).

My main concern is getting some data that reflects real workloads.  @jorisvandenbossche it looks it sounded like you had Geo data that has multiple levels of nesting, I wonder if there is a canonical dataset we could make use of for benchmarking.

pitrou reviewed (approved):
+1
